File: backend/app/models/attendance.py
from datetime import datetime, date
from sqlalchemy import Column, Integer, String, Date, DateTime, ForeignKey, UniqueConstraint
from sqlalchemy.orm import relationship

from app.database import Base


class Attendance(Base):
    """출석 기록"""
    __tablename__ = "attendances"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    date = Column(Date, nullable=False, index=True)
    status = Column(String(20), default='present')  # present, absent, recovered
    recovered_by_column_id = Column(Integer, ForeignKey("team_columns.id"), nullable=True)
    approved_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)

    __table_args__ = (
        UniqueConstraint('user_id', 'date', name='uq_user_date'),
    )

    # Relationships
    user = relationship("User", back_populates="attendances", foreign_keys=[user_id])
    approver = relationship("User", foreign_keys=[approved_by])
    recovery_column = relationship("TeamColumn", foreign_keys=[recovered_by_column_id])

File: spring-backend/backend/src/main/java/com/fundmessenger/attendance/entity/Attendance.java
package com.fundmessenger.attendance.entity;

import com.fundmessenger.column.entity.TeamColumn;
import com.fundmessenger.user.entity.User;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.OffsetDateTime;

@Entity
@Table(name = "attendances", uniqueConstraints = {
        @UniqueConstraint(name = "uq_user_date", columnNames = {"user_id", "date"})
})
@Getter
@Setter
@NoArgsConstructor
public class Attendance {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Column(name = "user_id", nullable = false, insertable = false, updatable = false)
    private Long userId;

    @Column(name = "date", nullable = false)
    private LocalDate date;

    @Column(name = "status", length = 20, columnDefinition = "varchar(20) default 'present'")
    private String status = "present";

    @Column(name = "recovered_by_column_id", insertable = false, updatable = false)
    private Long recoveredByColumnId;

    @Column(name = "approved_by", insertable = false, updatable = false)
    private Long approvedById;

    @Column(name = "created_at", columnDefinition = "timestamptz")
    private OffsetDateTime createdAt;

    // --- ManyToOne Relationships ---

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "approved_by")
    private User approver;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "recovered_by_column_id")
    private TeamColumn recoveryColumn;
}
